services/src/test/scala/cromwell/pipeline/service/impls/ProjectFileServiceIdCombinedTestImpl.scala

```diff
@@ -0,0 +1,73 @@
+package cromwell.pipeline.service.impls
+
+import cromwell.pipeline.datastorage.dto._
+import cromwell.pipeline.model.wrapper.UserId
+import cromwell.pipeline.service.{ ProjectFileService, VersioningException }
+
+import java.nio.file.Path
+import scala.concurrent.Future
+
+class ProjectFileServiceIdCombinedTestImpl(projectFileBundles: Seq[(ProjectId, Seq[ProjectFile])], testMode: TestMode)
+    extends ProjectFileService {
+
+  override def validateFile(fileContent: ProjectFileContent): Future[Either[ValidationError, Unit]] =
+    testMode match {
+      case WithException(exc: ValidationError) => Future.successful(Left(exc))
+      case WithException(exc)                  => Future.failed(exc)
+      case _                                   => Future.successful(Right(()))
+    }
+
+  override def uploadFile(
+    projectId: ProjectId,
+    projectFile: ProjectFile,
+    version: Option[PipelineVersion],
+    userId: UserId
+  ): Future[Either[VersioningException, Unit]] =
+    testMode match {
+      case WithException(exc: VersioningException) => Future.successful(Left(exc))
+      case WithException(exc)                      => Future.failed(exc)
+      case _                                       => Future.successful(Right(()))
+    }
+
+  override def getFile(
+    projectId: ProjectId,
+    path: Path,
+    version: Option[PipelineVersion],
+    userId: UserId
+  ): Future[ProjectFile] =
+    testMode match {
+      case WithException(exc) => Future.failed(exc)
+      case _ =>
+        val filesOpt = projectFileBundles.find {
+          case (id, _) => id == projectId
+        }.map(_._2)
+        val file = filesOpt.flatMap(seq => seq.find(_.path == path)).get
+
+        Future.successful(file)
+    }
+
+  override def getFiles(
+    projectId: ProjectId,
+    version: Option[PipelineVersion],
+    userId: UserId
+  ): Future[List[ProjectFile]] =
+    testMode match {
+      case WithException(exc) => Future.failed(exc)
+      case _ =>
+        val files = projectFileBundles.find {
+          case (id, _) => id == projectId
+        }.map(_._2).get.toList
```

Comment:
What if it's not found? Aren't you supposed to just return an empty list? Do you really need to call `get` on `Option` in this case?

Reply:
Method recreated